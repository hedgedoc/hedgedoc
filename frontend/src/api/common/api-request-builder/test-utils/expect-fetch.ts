/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { defaultConfig } from '../../default-config'
import { Mock } from 'ts-mockery'

/**
 * Mock fetch api for tests.
 * Check that the given url and options are present in the request and return the given status code.
 *
 * @param expectedUrl the url that should be requested
 * @param requestStatusCode the status code the mocked request should return
 * @param expectedOptions additional options
 */
export const expectFetch = (
  expectedUrl: string,
  requestStatusCode: number,
  expectedOptions: RequestInit,
  responseBody?: unknown
): void => {
  global.fetch = jest.fn((fetchUrl: RequestInfo | URL, fetchOptions?: RequestInit): Promise<Response> => {
    expect(fetchUrl).toEqual(expectedUrl)
    expect(fetchOptions).toStrictEqual({
      ...defaultConfig,
      body: undefined,
      headers: new Headers(),
      ...expectedOptions
    })
    return Promise.resolve(
      Mock.of<Response>({
        status: requestStatusCode,
        statusText: mapCodeToText(requestStatusCode),
        json: jest.fn(() => (responseBody ? Promise.resolve(responseBody) : Promise.reject()))
      })
    )
  }) as typeof global.fetch
}
const mapCodeToText = (code: number): string => {
  switch (code) {
    case 400:
      return 'bad_request'
    case 401:
      return 'forbidden'
    default:
      return 'unknown_code'
  }
}
