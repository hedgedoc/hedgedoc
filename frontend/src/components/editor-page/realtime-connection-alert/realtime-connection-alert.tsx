/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { UiIcon } from '../../common/icons/ui-icon'
import React from 'react'
import { Alert } from 'react-bootstrap'
import { ArrowRepeat as IconArrowRepeat } from 'react-bootstrap-icons'
import { Trans, useTranslation } from 'react-i18next'

/**
 * Modal with a spinner that is only shown while reconnecting to the realtime backend
 */
export const RealtimeConnectionAlert: React.FC = () => {
  useTranslation()

  return (
    <Alert variant={'warning'} className={'w-100 m-0 px-2 py-1 border-top-0 border-bottom-0 d-flex align-items-center'}>
      <Trans i18nKey={'realtime.connecting'}></Trans>
      <UiIcon icon={IconArrowRepeat} spin={true} className={'ms-2 mb-1'} />
    </Alert>
  )
}
