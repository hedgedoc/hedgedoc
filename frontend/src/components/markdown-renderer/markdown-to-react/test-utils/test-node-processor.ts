/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { Document } from 'domhandler'
import { Element, Text } from 'domhandler'

export class TestNodeProcessor extends NodeProcessor {
  process(document: Document): Document {
    document.childNodes.push(new Element('node-processor', {}, [new Text('node processor children')]))
    return document
  }
}
