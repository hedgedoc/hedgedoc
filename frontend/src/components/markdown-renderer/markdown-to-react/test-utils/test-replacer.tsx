/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { NativeRenderer, NodeReplacement, SubNodeTransform } from '../../replace-components/component-replacer'
import { ComponentReplacer, DO_NOT_REPLACE } from '../../replace-components/component-replacer'
import type { Element } from 'domhandler'
import React, { Fragment } from 'react'

export class TestReplacer extends ComponentReplacer {
  replace(node: Element, subNodeTransform: SubNodeTransform, nativeRenderer: NativeRenderer): NodeReplacement {
    return node.tagName === 'node-processor' ? (
      <Fragment>
        <span>NodeProcessor! </span>
        <span data-children={true}> {node.childNodes.map(subNodeTransform)} </span>
        <span data-native={true}> {nativeRenderer()} </span>
      </Fragment>
    ) : (
      DO_NOT_REPLACE
    )
  }
}
