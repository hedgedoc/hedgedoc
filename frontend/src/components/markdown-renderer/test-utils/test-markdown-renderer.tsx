/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { StoreProvider } from '../../../redux/store-provider'
import type { MarkdownRendererExtension } from '../extensions/_base-classes/markdown-renderer-extension'
import { MarkdownToReact } from '../markdown-to-react/markdown-to-react'
import React, { useMemo } from 'react'

export interface SimpleMarkdownRendererProps {
  content: string
  extensions: MarkdownRendererExtension[]
}

/**
 * A markdown renderer for tests.
 *
 * @param content The content to be rendered.
 * @param extensions The {@link MarkdownRendererExtension MarkdownExtensions} to use for rendering.
 */
export const TestMarkdownRenderer: React.FC<SimpleMarkdownRendererProps> = ({ content, extensions }) => {
  const lines = useMemo(() => content.split('\n'), [content])
  return (
    <StoreProvider>
      <MarkdownToReact
        markdownContentLines={lines}
        markdownRenderExtensions={extensions}
        newlinesAreBreaks={true}
        allowHtml={false}
      />
    </StoreProvider>
  )
}
