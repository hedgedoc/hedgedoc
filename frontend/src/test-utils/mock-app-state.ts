/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import * as useApplicationStateModule from '../hooks/common/use-application-state'
import type { ApplicationState } from '../redux'
import { initialState as initialStateDarkMode } from '../redux/dark-mode/initial-state'
import { initialState as initialStateEditorConfig } from '../redux/editor-config/initial-state'
import { initialState as initialStateNoteDetails } from '../redux/note-details/initial-state'
import { initialState as initialStateRealtimeStatus } from '../redux/realtime/initial-state'
import { initialState as initialStateRendererStatus } from '../redux/renderer-status/initial-state'
import type { NoteDetails } from '../redux/note-details/types'
import type { RealtimeStatus } from '../redux/realtime/types'
import type { DeepPartial } from '@hedgedoc/commons'

jest.mock('../redux/editor-config/methods', () => ({
  loadFromLocalStorage: jest.fn().mockReturnValue(undefined)
}))
jest.mock('../hooks/common/use-application-state')

/**
 * Mocks the {@link ApplicationState} for a test.
 * When not overriden, it uses the initial state of the reducers.
 *
 * @param state Overrides for the mocked state
 */
export const mockAppState = (state?: DeepPartial<ApplicationState>) => {
  jest.spyOn(useApplicationStateModule, 'useApplicationState').mockImplementation((fn) => {
    return fn({
      darkMode: {
        ...initialStateDarkMode,
        ...state?.darkMode
      },
      editorConfig: {
        ...initialStateEditorConfig,
        ...state?.editorConfig
      },
      history: state?.history ?? [],
      noteDetails: {
        ...initialStateNoteDetails,
        ...state?.noteDetails
      } as NoteDetails,
      realtimeStatus: {
        ...initialStateRealtimeStatus,
        ...state?.realtimeStatus
      } as RealtimeStatus,
      rendererStatus: {
        ...initialStateRendererStatus,
        ...state?.rendererStatus
      },
      user: state?.user ?? null
    })
  })
}
