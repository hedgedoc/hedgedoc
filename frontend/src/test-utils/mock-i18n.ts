/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { TFunction } from 'i18next'
import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'

/**
 * Initializes i18n with minimal settings and without any data, so it just returns the used key as translation.
 *
 * @return A promise that resolves if i18n has been initialized
 */
export const mockI18n = (): Promise<TFunction> => {
  return i18n.use(initReactI18next).init({
    lng: 'en',
    fallbackLng: 'en',
    ns: ['translationsNS'],
    defaultNS: 'translationsNS',
    interpolation: {
      escapeValue: false
    },
    resources: { en: { translationsNS: {} } }
  })
}
