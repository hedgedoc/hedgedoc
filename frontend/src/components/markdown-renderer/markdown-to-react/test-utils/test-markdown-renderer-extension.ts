/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { MarkdownRendererExtension } from '../../extensions/_base-classes/markdown-renderer-extension'
import type { NodeProcessor } from '../../node-preprocessors/node-processor'
import type { ComponentReplacer } from '../../replace-components/component-replacer'
import { TestNodeProcessor } from './test-node-processor'
import { TestReplacer } from './test-replacer'
import type MarkdownIt from 'markdown-it'
import Token from 'markdown-it/lib/token'

export class TestMarkdownRendererExtension extends MarkdownRendererExtension {
  buildNodeProcessors(): NodeProcessor[] {
    return [new TestNodeProcessor()]
  }

  configureMarkdownIt(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('configure', (core) => core.tokens.push(new Token('configure', 'configure', 0)))
      markdownIt.renderer.rules.configure = () => '<span>configure</span>'
    })
  }

  buildReplacers(): ComponentReplacer[] {
    return [new TestReplacer()]
  }

  buildTagNameAllowList(): string[] {
    return ['node-processor']
  }

  configureMarkdownItPost(markdownIt: MarkdownIt) {
    markdownIt.use(() => {
      markdownIt.core.ruler.push('post', (core) => core.tokens.push(new Token('post', 'post', 0)))
      markdownIt.renderer.rules.post = () => '<span>post</span>'
    })
  }
}
