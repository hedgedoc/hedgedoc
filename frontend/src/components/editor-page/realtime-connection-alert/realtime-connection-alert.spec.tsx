/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { mockI18n } from '../../../test-utils/mock-i18n'
import { RealtimeConnectionAlert } from './realtime-connection-alert'
import { render } from '@testing-library/react'

describe('realtime connection alert', () => {
  beforeAll(mockI18n)

  it('will render correctly', () => {
    const view = render(<RealtimeConnectionAlert />)
    expect(view.container).toMatchSnapshot()
  })
})
