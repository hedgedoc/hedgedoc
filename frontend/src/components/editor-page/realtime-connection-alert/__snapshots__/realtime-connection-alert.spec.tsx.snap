// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`realtime connection alert will render correctly 1`] = `
<div>
  <div
    class="fade w-100 m-0 px-2 py-1 border-top-0 border-bottom-0 d-flex align-items-center alert alert-warning show"
    role="alert"
  >
    realtime.connecting
    BootstrapIconMock_ArrowRepeat
  </div>
</div>
`;
